src/main/scala/org/apache/spark/shuffle/ucx/ShuffleTransport.scala

```diff
@@ -0,0 +1,126 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION. ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
+package org.apache.spark.shuffle.ucx
+
+import java.nio.ByteBuffer
+
+/**
+ * Class that represents some block in memory with it's address, size.
+ *
+ * @param isHostMemory host or GPU memory
+ */
+case class MemoryBlock(address: Long, size: Long, isHostMemory: Boolean = true)
+
+/**
+ * Opaque object to describe remote memory block (address, rkey, etc.).
+ */
+trait Cookie
+
+/**
+ * Base class to indicate some blockId. It should be hashable and could be constructed on both ends.
+ * E.g. ShuffleBlockId(shuffleId: Int, mapId: Long, reduceId: Int)
+ */
+trait BlockId
+
+trait Block {
+  // Transport will call this method when it would need an actual block memory.
+  def getMemoryBlock: MemoryBlock
+}
+
+
+object OperationStatus extends Enumeration {
+  val SUCCESS, FAILURE = Value
+}
+
+/**
+ * Operation statistic, like completionTime, transport used, protocol used, etc.
+ */
+trait OperationStats {
+  def getElapsedTimeNs: Long
+}
+
+class TransportError(errorMsg: String) extends Exception(errorMsg)
+
+trait OperationResult {
+  def recvSize: Long
+  def getStatus: OperationStatus.Value
+  def getError: TransportError
+  def getStats: OperationStats
+}
+
+/**
+ * Async operation callbacks
+ */
+trait OperationCallback {
+  def onComplete(result: OperationResult)
+}
+
+/**
+ * Transport flow example:
+ * val transport = new UcxShuffleTransport()
+ * transport.init()
+ *
+ * Mapper/writer:
+ * transport.register(metadataBlockId, metadataBlock) // we don't care for metadata cookies
+ *
+ * Reducer:
+ * transport.fetchBlockByBlockId(blockId, resultBounceBuffer)
+ *
+ * transport.unregister(blockId)
+ * transport.close()
+ */
+trait ShuffleTransport {
+
+  /**
+   * Initialize transport resources. This function should get called after ensuring that SparkConf
+   * has the correct configurations since it will use the spark configuration to configure itself.
+   * @return worker address of current process, to use in [[ addExecutor()]]
+   */
+  def init(): ByteBuffer
+
+  /**
+   * Close all transport resources
+   */
+  def close()
+
+  /**
+   * Add executor's worker address. For standalone testing purpose and for implementations that makes
+   * connection establishment outside of UcxShuffleManager.
+   */
+  def addExecutor(executorId: String, workerAddress: ByteBuffer)
```

Comment:
should also have `removeExecutor(executorId)` as well, since executors can go away, even for legitimate reasons.